### PR TITLE
Add the filesystem integrity module and test

### PR DESCRIPTION
**Pull request by roronoasins**
|Related issue|
|---|
|#2231|

## Description

In order to accomplish the need to check if an installation, upgrade, or uninstallation of the Wazuh environment performs
any unwanted changes in the filesystem, this PR adds a new test. With this new type of test, we can expand its behavior
to detect any Wazuh unwanted changes within its own files, that logs are not truncated by Wazuh, etc.

After research and study about the paths that change when performing a manager installation, uninstallation, or upgrade
the warning list has been updated. There are many files/paths that should not change, so the test is in red state for now
for all the current cases.

The changes within `/var/lib` and `/var/log` still remaining, a sync up is remaining to decide if they are warning or
red.

### Added

- Add `check_files.py` script to `wazuh_testing` framework
- Add `filesystem_integrity` test module in system tests
- Add `test_check_file_system_integrity` using `qa-ctl` tasks
- Add `filesystem_integrity/launcher.py` which runs qa-ctl and check the differences between check-files

## Checks

- [x] Works as expected in Ubuntu Focal
- [x] Works as expected in CentOS 7
- [x] Works as expected in CentOS 8

## Log example

<details><summary>Sample run</summary>
<p>

```bash
$ python3 launcher.py --qa-branch 2312-warning-list --output-file-path /tmp/syscheck/ubuntu/install -a install -o ubuntu
2022-01-07 15:35:09,883 - INFO - Validating input parameters
2022-01-07 15:35:10,130 - INFO - Input parameters validation has passed successfully
2022-01-07 15:35:10,712 - INFO - Generating qa-ctl configuration
2022-01-07 15:35:10,714 - INFO - The qa-ctl configuration has been created successfully in /tmp/wazuh_check_files/check_files_config_1641566110_712778.yaml
2022-01-07 15:35:12,547 - INFO - Starting 1 instances deployment
2022-01-07 15:37:08,273 - INFO - The instances deployment has finished sucessfully
2022-01-07 15:37:08,290 - INFO - Running download_check_files_tool task from playbook /tmp/wazuh_playbooks/download_files_1641566110_708031.yaml
2022-01-07 15:37:10,944 - INFO - The download_check_files_tool task from playbook /tmp/wazuh_playbooks/download_files_1641566110_708031.yaml has been finished successfully
2022-01-07 15:37:10,944 - INFO - Running run_pre_check_files_scan task from playbook /tmp/wazuh_playbooks/run_linux_commands_1641566110_70899.yaml
2022-01-07 15:43:32,417 - INFO - The run_pre_check_files_scan task from playbook /tmp/wazuh_playbooks/run_linux_commands_1641566110_70899.yaml has been finished successfully
2022-01-07 15:43:32,417 - INFO - Running install_wazuh task from playbook /tmp/wazuh_playbooks/install_wazuh_1641566110_709608.yaml
2022-01-07 15:45:05,247 - INFO - The install_wazuh task from playbook /tmp/wazuh_playbooks/install_wazuh_1641566110_709608.yaml has been finished successfully
2022-01-07 15:45:05,247 - INFO - Running waiting_time_before_running_check_files_tool task from playbook /tmp/wazuh_playbooks/wait_seconds_1641566110_710865.yaml
2022-01-07 15:45:35,911 - INFO - The waiting_time_before_running_check_files_tool task from playbook /tmp/wazuh_playbooks/wait_seconds_1641566110_710865.yaml has been finished successfully
2022-01-07 15:45:35,912 - INFO - Running run_post_check_files_scan task from playbook /tmp/wazuh_playbooks/run_linux_commands_1641566110_711265.yaml
2022-01-07 15:51:11,533 - INFO - The run_post_check_files_scan task from playbook /tmp/wazuh_playbooks/run_linux_commands_1641566110_711265.yaml has been finished successfully
2022-01-07 15:51:11,533 - INFO - Running fetch check-files data task from playbook /tmp/wazuh_playbooks/fetch_files_1641566110_711866.yaml
2022-01-07 15:52:25,175 - INFO - The fetch check-files data task from playbook /tmp/wazuh_playbooks/fetch_files_1641566110_711866.yaml has been finished successfully
2022-01-07 15:52:25,200 - INFO - The post-check-files data has been saved in /tmp/wazuh_check_files/post_check_files_data_1641566107_47861.json
2022-01-07 15:52:25,200 - INFO - The pre-check-files data has been saved in /tmp/wazuh_check_files/pre_check_files_data_1641566107_47861.json
2022-01-07 15:53:10,604 - INFO - The test results have been stored in /tmp/syscheck/ubuntu/install
============================= test session starts ==============================
platform linux -- Python 3.8.10, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /tmp
plugins: testinfra-5.0.0, metadata-1.11.0, html-2.0.1
collected 1 item

test_check_file_system_integrity.py F                                    [100%]

=================================== FAILURES ===================================
_______________________ test_check_file_system_integrity _______________________

get_first_file = '/tmp/wazuh_check_files/pre_check_files_data_1641566107_47861.json'
get_second_file = '/tmp/wazuh_check_files/post_check_files_data_1641566107_47861.json'
get_output_path = '/tmp/syscheck/ubuntu/install'

    def test_check_file_system_integrity(get_first_file, get_second_file, get_output_path):
        """This test checks if two files are not equal and if differences exist, and check if the path where are changes
        is listed in the warning list or not.
    
        After an installation, update or uninstallation is necessary to check if the system files are the same as before.
        Two given files are checked for differences between them. After checking their contents, if there are differences
        between them, the test will check if any of the paths that have changed is listed in the warning list(which would
        change the test state to 'warning', aka 'yellow') or not(which would change the test state to 'failed', aka 'red').
        These differences(a file for each state) will be stored in the path that the user passes as an argument.
    
        If the test shows a warning related to the differences, the state of the test exectuion requires a manual revision.
    
        The possible states after the test execution are:
            - yellow: When there are paths with changes but they appear in the warning list.
            - red: When there are paths with changes that do not appear in the warning list.
    
        Example run:
            python3 -m pytest wazuh-qa/tests/check_files/test_check_files/test_system_check_files.py
            --before-file initial_state --after_file after_installing_manager --output-path /tmp/system_check_files
    
        Args:
            get_first_file (fixture): Get the file before making any changes to the environment.
            get_second_file (fixture): Get the file after making any changes to the environment.
            get_output_path (fixture): Get the output path where the result will be saved.
        """
        file1_data = validate_and_read_json(get_first_file)
        file2_data = validate_and_read_json(get_second_file)
        validate_and_create_output_path(get_output_path)
    
        # The DeepDiff module gives us the differences between these two files.
        differences = DeepDiff(file1_data, file2_data)
        differences_str = differences.to_json().replace('root', '')
    
        # If there are differences between the given files
        if differences != {}:
            differences_json = json.loads(differences_str)
            test_state, yellos_json_dict, red_json_dict = check_diffs_in_warning_list(differences_json,
                                                                                      read_warning_list(WARNING_LIST_PATH))
    
            yellow_path = os.path.join(get_output_path, 'warning_diff.json')
            write_json_file(yellow_path, yellos_json_dict)
            warnings.warn("There are some directories that are contained in the warning list. "
                          f"Please check {yellow_path} file in order to determinate if the test passes "
                          "or not.", UserWarning)
    
            if test_state == 'red':
                red_path = os.path.join(get_output_path, 'fail_diff.json')
                write_json_file(red_path, red_json_dict)
>               raise AssertionError("There are some directories that not contained within the warning list. "
                                     f"These paths are logged here: {red_path}.")
E               AssertionError: There are some directories that not contained within the warning list. These paths are logged here: /tmp/syscheck/ubuntu/install/fail_diff.json.

test_check_file_system_integrity.py:313: AssertionError
=============================== warnings summary ===============================
wazuh_check_files/wazuh-qa/tests/check_file_system_integrity/test_check_file_system_integrity.py::test_check_file_system_integrity
  /tmp/wazuh_check_files/wazuh-qa/tests/check_file_system_integrity/test_check_file_system_integrity.py:306: UserWarning: There are some directories that are contained in the warning list. Please check /tmp/syscheck/ubuntu/install/warning_diff.json file in order to determinate if the test passes or not.
    warnings.warn("There are some directories that are contained in the warning list. "

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED test_check_file_system_integrity.py::test_check_file_system_integrity
======================== 1 failed, 1 warning in 43.55s =========================

AssertionError: The check file system integrity test has failed

```

</p>
</details>

<details><summary><code>fail_diff.json</code></summary>
<p>

```json
{
    "dictionary_item_added": [
        "/usr/share/lintian/overrides/wazuh-manager",
        "/usr/local/lib/python3.9/dist-packages/apt/__pycache__",
        "/usr/local/lib/python3.9/dist-packages/apt/__pycache__/__init__.cpython-39.pyc",
        "/usr/local/lib/python3.9/dist-packages/apt/__pycache__/cache.cpython-39.pyc",
        "/usr/local/lib/python3.9/dist-packages/apt/__pycache__/debfile.cpython-39.pyc",
        "/usr/local/lib/python3.9/dist-packages/apt/__pycache__/cdrom.cpython-39.pyc",
        "/usr/local/lib/python3.9/dist-packages/apt/__pycache__/package.cpython-39.pyc",
        "/usr/local/lib/python3.9/dist-packages/apt/progress/__pycache__",
        "/usr/local/lib/python3.9/dist-packages/apt/progress/__pycache__/base.cpython-39.pyc",
        "/usr/local/lib/python3.9/dist-packages/apt/progress/__pycache__/text.cpython-39.pyc",
        "/usr/local/lib/python3.9/dist-packages/apt/progress/__pycache__/__init__.cpython-39.pyc",
        "/tmp"
    ],
    "values_changed": {
        "/usr/lib/modules/5.4.0-86-generic": {
            "last_update": {
                "new_value": "2022-01-07 14:39:16",
                "old_value": "2021-10-05 09:42:29"
            }
        },
        "/usr/lib/modules/5.4.0-86-generic/modules.dep": {
            "last_update": {
                "new_value": "2022-01-07 14:39:16",
                "old_value": "2021-10-05 09:42:28"
            }
        },
        "/usr/lib/modules/5.4.0-86-generic/modules.dep.bin": {
            "last_update": {
                "new_value": "2022-01-07 14:39:16",
                "old_value": "2021-10-05 09:42:28"
            }
        },
        "/usr/lib/modules/5.4.0-86-generic/modules.symbols.bin": {
            "last_update": {
                "new_value": "2022-01-07 14:39:16",
                "old_value": "2021-10-05 09:42:29"
            }
        },
        "/usr/lib/modules/5.4.0-86-generic/modules.softdep": {
            "last_update": {
                "new_value": "2022-01-07 14:39:16",
                "old_value": "2021-10-05 09:42:29"
            }
        },
        "/usr/lib/modules/5.4.0-86-generic/modules.builtin.bin": {
            "last_update": {
                "new_value": "2022-01-07 14:39:16",
                "old_value": "2021-10-05 09:42:29"
            }
        },
        "/usr/lib/modules/5.4.0-86-generic/modules.devname": {
            "last_update": {
                "new_value": "2022-01-07 14:39:16",
                "old_value": "2021-10-05 09:42:29"
            }
        },
        "/usr/lib/modules/5.4.0-86-generic/modules.symbols": {
            "last_update": {
                "new_value": "2022-01-07 14:39:16",
                "old_value": "2021-10-05 09:42:29"
            }
        },
        "/usr/lib/modules/5.4.0-86-generic/modules.builtin.alias.bin": {
            "last_update": {
                "new_value": "2022-01-07 14:39:16",
                "old_value": "2021-10-05 09:42:29"
            }
        },
        "/usr/lib/modules/5.4.0-86-generic/modules.alias.bin": {
            "last_update": {
                "new_value": "2022-01-07 14:39:16",
                "old_value": "2021-10-05 09:42:29"
            }
        },
        "/usr/lib/modules/5.4.0-86-generic/modules.alias": {
            "last_update": {
                "new_value": "2022-01-07 14:39:16",
                "old_value": "2021-10-05 09:42:28"
            }
        },
        "/usr/share/lintian/overrides": {
            "last_update": {
                "new_value": "2022-01-07 14:44:11",
                "old_value": "2021-10-05 09:51:32"
            }
        },
        "/usr/local/lib/python3.9/dist-packages/apt": {
            "last_update": {
                "new_value": "2022-01-07 14:43:47",
                "old_value": "2021-10-05 10:02:48"
            }
        },
        "/usr/local/lib/python3.9/dist-packages/apt/progress": {
            "last_update": {
                "new_value": "2022-01-07 14:43:47",
                "old_value": "2021-10-05 10:02:48"
            }
        },
        "/etc/mtab": {
            "last_update": {
                "new_value": "2022-01-07 14:49:56",
                "old_value": "2022-01-07 14:42:16"
            },
            "md5sum": {
                "new_value": "4eff20d83a2e99bd0972f70a2dc83e61",
                "old_value": "5664471f37cd400a4cc99ac216250a72"
            }
        }
    }
}
```

</p>
</details>

## Results when target is manager

<table>
<thead>
<tr>
<th style="width: 175px;">Name</th>
<th style="width: 499px;" colspan="6">Local</th>
</tr>
</thead>
<tbody>
<tr>
<td style="width: 175px;">OS</td>
<td style="width: 208px;" colspan="3">Linux</td>
<td style="width: 208px;" colspan="3">Windows</td>
</tr>
<tr>
<td style="width: 175px;">Action</td>
<td style="width: 103px;">Install Wazuh</td>
<td style="width: 99px;">Uninstall Wazuh</td>
<td style="width: 79px;">Upgrade Wazuh</td>
<td style="width: 103px;">Install Wazuh</td>
<td style="width: 99px;">Uninstall Wazuh</td>
<td style="width: 79px;">Upgrade Wazuh</td>
</tr>
<tr>
<td style="width: 175px;"><strong>Ubuntu Focal</strong></td>
<td style="width: 103px;">🔴</td>
<td style="width: 99px;">🔴</td>
<td style="width: 79px;">🔴</td>
<td style="width: 103px;">N/A</td>
<td style="width: 99px;">N/A</td>
<td style="width: 79px;">N/A</td>
</tr>
<tr> 
<td style="width: 175px;"><strong>CentOS7</strong></td>
<td style="width: 103px;">🔴</td>
<td style="width: 99px;">🔴</td>
<td style="width: 79px;">🔴</td>
<td style="width: 103px;">N/A</td>
<td style="width: 99px;">N/A</td>
<td style="width: 79px;">N/A</td>
</tr>
<tr>
<td style="width: 175px;"><strong>CentOS8</strong></td>
<td style="width: 103px;">🔴</td>
<td style="width: 99px;">🔴</td>
<td style="width: 79px;">🔴</td>
<td style="width: 103px;">N/A</td>
<td style="width: 99px;">N/A</td>
<td style="width: 79px;">N/A</td>
</tr>

</tbody>
</table>

## Tests

- [x] Proven that tests **fail** when they have to fail.
- [x] Python codebase satisfies PEP-8 style style guide. `pycodestyle --max-line-length=120 --show-source --show-pep8 file.py`.
- [x] Python codebase is documented following the Google Style for Python docstrings.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -42,5 +42,6 @@ bandit==1.7.0
 git-repo==1.10.3
 pywinrm>=0.4.2 ; platform_system == "Linux" or platform_system=='Windows'
 wmi>=1.5.1; platform_system=='Windows'
+deepdiff
```
Comment (jmv74211) [inline requirements.txt:45]: Add version and system pls

File: tests/system/filesystem_integrity/README.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
+# Filesystem integrity test
```
Comment (jmv74211) [inline tests/system/filesystem_integrity/README.md:1]: Have you tried to run the test according to this documentation?

File: tests/system/filesystem_integrity/conftest.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# Copyright (C) 2015-2021, Wazuh Inc.
+# Created by Wazuh, Inc. <info@wazuh.com>.
+# This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
+
```
Comment (jmv74211) [inline tests/system/filesystem_integrity/conftest.py:4]: We have to move the `filesystem_integrity` from system tests to outside.

File: tests/system/filesystem_integrity/data/warning_list.txt [hunk 4]
```diff
@@ -0,0 +1,31 @@
+/etc/group
```
Comment (jmv74211) [inline tests/system/filesystem_integrity/data/warning_list.txt:1]: Add this file to `data/checkfiles/warning_list.txt`. Check that when the framework is installed, this file is placed in this path too

**Review by jmv74211 — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 1 above)
Comment (roronoasins) [reply requirements.txt]: Solved https://github.com/wazuh/wazuh-qa/pull/2437/commits/7955e10644d7e0d5c4e2db1e82d28b85e367946a

File: tests/system/filesystem_integrity/README.md (hunk 2 above)
Comment (roronoasins) [reply tests/system/filesystem_integrity/README.md]: :+1:

File: tests/system/filesystem_integrity/conftest.py (hunk 3 above)
Comment (roronoasins) [reply tests/system/filesystem_integrity/conftest.py]: Solved https://github.com/wazuh/wazuh-qa/pull/2437/commits/7955e10644d7e0d5c4e2db1e82d28b85e367946a

File: tests/system/filesystem_integrity/data/warning_list.txt (hunk 4 above)
Comment (roronoasins) [reply tests/system/filesystem_integrity/data/warning_list.txt]: Solved https://github.com/wazuh/wazuh-qa/pull/2437/commits/7955e10644d7e0d5c4e2db1e82d28b85e367946a

File: tests/check_files/test_filesystem_integrity/data/warning_list.txt [hunk 5]
```diff
@@ -0,0 +1,31 @@
+/etc/group
```
Comment (jmv74211) [inline tests/check_files/test_filesystem_integrity/data/warning_list.txt:1]: You are duplicating this file.

You have to decide if add this file to the framework data files, or as testing data file,

**Review by jmv74211 — CHANGES_REQUESTED** [review]

File: tests/check_files/test_filesystem_integrity/data/warning_list.txt (hunk 5 above)
Comment (roronoasins) [reply tests/check_files/test_filesystem_integrity/data/warning_list.txt:1]: Solved https://github.com/wazuh/wazuh-qa/pull/2437/commits/24801f68dd74da1dce8fb303004cbb0c65e79a25

I think it is better to have it with all the check-files tests

**Review by jmv74211 — APPROVED** [review]

**Pull request merged**